File: day1/part2.go
package main

import (
    "fmt"
    "log"
    "os"
    "bufio"
    "strings"
    "strconv"
)

func main()  {
    file, err := os.Open("input.txt")
    if err != nil {
        log.Fatal("Failed to open file: %v", err)
    }
    defer file.Close()

    r := bufio.NewReader(file)
    var list1 []int
    var list2 []int
    for {
        line, _, err := r.ReadLine()
        if err != nil {
            break
        }

        numbers := strings.Split(string(line), "   ")
        n1, _ := strconv.Atoi(numbers[0])
        list1 = append(list1, n1)

        n2, _ := strconv.Atoi(numbers[1])
        list2 = append(list2, n2)
    }

    m := make(map[int]int)
    for _, v := range list2 {
        m[v]++
    }

    sum := 0
    for _, v := range list1 {
        sum += v * m[v]
    }
    fmt.Println(sum)

}

File: day2/part2.py
numbers = []

def is_safe(numbers):
    max = 3
    is_increasing = numbers[0] < numbers[1]
    for i in range(0, len(numbers) - 1):
        if is_increasing and numbers[i] >= numbers[i + 1]:
            return False
        if is_increasing and numbers[i + 1] - numbers[i] > max:
            return False
        if not is_increasing and numbers[i] <= numbers[i + 1]:
            return False
        if not is_increasing and numbers[i] - numbers[i + 1]  > max:
            return False

    return True

safe_reports = 0
with open("input.txt") as file:
    report = -1
    for line in file:
        numbers = [int(n) for n in line.split(" ")]
        safe = is_safe(numbers)
        if safe:
            safe_reports+=1
            continue

        # try each variant
        for i in range(0, len(numbers)):
            modified_numbers = numbers[0:i] + numbers[i+1:len(numbers)]
            safe = is_safe(modified_numbers)
            if safe:
                safe_reports+=1
                break

print(f'safe reports: {safe_reports}')